share/wake/lib/core/list.wake

```diff
@@ -67,14 +67,26 @@ export def tail: List a => List a = match _
     Nil  = Nil
     _, t = t
 
-# map: create a new List by applying a function f to each element of a List.
-#
-#   map str    (seq 5) = "0", "1", "2", "3", "4", Nil
-#   map (_+10) (seq 5) = 10, 11, 12, 13, 14, Nil
-export def map (f: a => b): List a => List b =
-    def loop = match _
+# Create a new List by applying the function *mapFn* to each element of *list*.
+# The *map* function (along with *foldl*) is generally how one implements loops in wake.
```

Comment:
One style convention that I see often in other languages is to put a blank line between the first line/sentence and subsequent sentence. The style also involves making sure that first sentence is a good, single-line summary sentence.

I don't know what it'll look like in VSCode, since I don't use it, but many documentation site generators will often display only that first line in scenarios where real estate is limited. For example, in Scala, [only that first line is displayed](https://www.scala-lang.org/api/2.13.3/scala/collection/immutable/List.html) in the list of all methods, and you have to click on the method to see the longer description.

Here's examples of language style guides for the separate summary line:

- Scala: https://docs.scala-lang.org/style/scaladoc.html
- Python: https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings
- Doxygen (many languages, but probably mostly C++): https://www.doxygen.nl/manual/docblocks.html#cppblock

Reply:
Here's what the special first-line style produces:

![Screen Shot 2021-09-14 at 6 14 15 PM](https://user-images.githubusercontent.com/1101706/133354361-d0f04650-4a29-4ff2-878b-8a6e648e6239.png)

Reply:
Having the first sentence be a succinct summary is a good idea.

However, I'm not sure we need to put a blank line in the comment between it and the next line. You can have a good first sentence that can be pulled out by hypothetical future tooling and still have more sentences that end up in the same paragraph if there's space.

I think I lean towards the formatting of the original proposal, but adding the requirement that the first sentence should be a self-contained description, suitable for extract.

Reply:
Maybe the rule is: the first line in the wake comment should be a complete self-contained sentence. Subsequent text that elaborates can follow on the next line to continue the paragraph.

Reply:
I like that formulation of the rule.  It also leads naturally into keeping that summary short, as too long of a summary will result in the first sentence taking up multiple lines.  The only thing I can see tripping it up is that, if we ever get code formatters with knowledge of comment blocks, they may want to update the line breaks in that paragraph to "better" justify the right edge; I don't think Vim's paragraph detection, at least, would be sophisticated enough to understand "ignore the first line of a comment paragraph if and only if it is the first paragraph in a documentation block".  That's probably not a big enough issue to block the guideline, though it might suggest leaning toward inserting a blank line.